### PR TITLE
feat(rest.packages.provider): add `deploy/v2` Rest API

### PR DESCRIPTION
**Pull request by mattdibi**
This PR adds the `deploy/v2` RestAPI endpoint to the Kura codebase. The endpoint was already available in the `ansible` bundle but it is now being promoted to core.

In its current form it is a 1:1 matching with the old implementation, but I will contribute additional functionalities in the next PRs.

**Roadmap**:
- [x] Contribute `deploy/v2` in its current form (#4868)
- [ ] Add endpoint for Deployment Package upload using `multipart/form-data`
- [ ] Add Depoloyment Package descriptor download and checking to `install` endpoint

**Changelog**:
- Added `org.eclipse.kura.rest.packages.provider` containing the implementation code
- Added `org.eclipse.kura.rest.packages.provider.test` containing both unit and integration tests

**Differences with the existing implementation**: differently from the `ansible` bundle implementation I had to:
- Complete re-write of the integration tests
- Refactored unit tests to adhere to Gerkhin style
- Change package names

---

## Tests

I ran the following tests on a Nvidia Jetson Nano running the latest buid with this PR

```bash
export ADDRESS=<addr>
export USERNAME=<user>
export PASSWORD=<pass>
```

then used `curl` to run the requests.

#### Get installed packages

```bash
curl -k -u $USERNAME:$PASSWORD https://$ADDRESS/services/deploy/v2/
[]%
```

after installing the [Wire Script Filter](https://marketplace.eclipse.org/content/wires-script-filter-kura-45)

```bash
curl -k -u $USERNAME:$PASSWORD https://$ADDRESS/services/deploy/v2/
[{"name":"org.eclipse.kura.wire.script.filter","version":"1.3.0"}]%
```

#### Uninstall a package

```bash
curl -X DELETE -k -u $USERNAME:$PASSWORD https://$ADDRESS/services/deploy/v2/org.eclipse.kura.wire.script.filter
"REQUEST_RECEIVED"%
```

The package is correctly uninstalled

#### Install package from URL

```bash
curl -X POST -k -u $USERNAME:$PASSWORD -H 'Content-Type: application/json' -d '{ "url": "https://download.eclipse.org/kura/releases/5.3.0/org.eclipse.kura.wire.script.filter-1.3.0.dp" }' https://$ADDRESS/services/deploy/v2/_install
```

The package is correctly installed (after updating the HTTPS certificate)

---

#### Get installed packages
- Method: GET
- API PATH: `/deploy/v2/`

##### Responses
- 200 OK status

```JSON
[{ "name": "packageName", "version": "packageVersion"}]
```

#### Install package from URL
- Method: POST
- API PATH: `/deploy/v2/_install`

##### Responses
- 200 OK status

```
"REQUEST_RECEIVED"
```

#### Uninstall a package
- Method: DELETE
- API PATH: `/deploy/v2/{name}`

##### Responses
- 200 OK status

```
"REQUEST_RECEIVED"
```

### CODE REVIEW

File: kura/org.eclipse.kura.rest.packages.provider/OSGI-INF/deployment_rest_service.xml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+	 Eurotech
+
+-->
+<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0" name="org.eclipse.kura.rest.deployment.agent.DeploymentRestService">
+   <implementation class="org.eclipse.kura.internal.rest.deployment.agent.DeploymentRestService"/>
+   <property name="kura.service.pid" type="String" value="org.eclipse.kura.rest.deployment.agent.DeploymentRestService"/>
+   <service>
+      <provide interface="org.eclipse.kura.internal.rest.deployment.agent.DeploymentRestService"/>
+   </service>
+   <reference bind="setDeploymentAgentService" cardinality="1..1" interface="org.eclipse.kura.deployment.agent.DeploymentAgentService" name="DeploymentAgentService" policy="static" unbind="unsetDeploymentAgentService"/>
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.rest.packages.provider/OSGI-INF/deployment_rest_service.xml]: Typically if the policy is static, the unbind method is not required.

File: kura/org.eclipse.kura.rest.packages.provider/OSGI-INF/deployment_rest_service.xml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+   <reference name="DeploymentAdmin"
+              policy="static"
+              cardinality="1..1"
+              bind="setDeploymentAdmin"
+              unbind="unsetDeploymentAdmin"
+              interface="org.osgi.service.deploymentadmin.DeploymentAdmin"/>
+   <reference bind="setUserAdmin" unbind="unsetUserAdmin" cardinality="0..1" interface="org.osgi.service.useradmin.UserAdmin" name="UserAdmin" policy="static"/>
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.rest.packages.provider/OSGI-INF/deployment_rest_service.xml]: Could you please format the references in the same way?

Comment (mattdibi) [reply kura/org.eclipse.kura.rest.packages.provider/OSGI-INF/deployment_rest_service.xml]: Fixed in 171aab713b8e7da97fa4264b20bc6d5f0d67c87f

File: kura/org.eclipse.kura.rest.packages.provider/OSGI-INF/deployment_rest_service.xml (hunk 1 above)
Comment (mattdibi) [reply kura/org.eclipse.kura.rest.packages.provider/OSGI-INF/deployment_rest_service.xml]: Fixed in fd6188a036de2fdbe67b4cb2c5d89f374edb26f8

**Review by pierantoniomerlino — APPROVED** [review]

**Pull request merged**